atl-checker/src/algorithms/global/mod.rs

```diff
@@ -0,0 +1,466 @@
+                    }
+                }
+            }
+        }
+    }
+
+    /// Inserts a vertex in the set at the index of curr_dist of dist.
+    fn insert_in_curr_dist(&mut self, v: V, dist: u32) {
+        match self.dist.get_mut(dist as usize) {
+            None => {
+                let mut curr_dist = HashSet::<V>::new();
+                curr_dist.insert(v);
+                self.dist.insert(dist as usize, curr_dist);
+            }
+            Some(set) => {
+                set.insert(v);
+            }
+        }
+    }
+
+    /// The resemblance of the function described in the paper, but in order to
+    /// know which component we are worker with and the assignments of the component
+    /// before, these are both given as arguments. The function it self returns a boolean
+    /// which is true if the assignments are changed.
+    fn f(&mut self, component: HashSet<V>, ass_from_earlier: HashMap<V, bool>) -> bool {
+        let mut changed_flag = false;
+
+        for vertex in component {
+            for edge in self.edg.succ(&vertex) {
+                match edge {
+                    Edge::Hyper(e) => changed_flag = max(self.process_hyper(e), changed_flag),
+                    Edge::Negation(e) => {
+                        changed_flag = max(
+                            self.process_negation(e, ass_from_earlier.clone()),
+                            changed_flag,
+                        )
+                    }
+                }
+            }
+        }
+        changed_flag
+    }
+
+    /// Rising the value of the source vertex assignment if all targets are true or empty. If the
+    /// source vertex already is true we simply return. The return value is based on if a changed
+    /// have been made.
+    fn process_hyper(&mut self, edge: HyperEdge<V>) -> bool {
+        if *self.assignment.get(&edge.source).unwrap() {
+            false
+        } else {
+            let mut final_ass = true;
+            for target in edge.targets {
+                if !self.assignment.get(&target).unwrap() {
+                    final_ass = false;
+                    break;
+                }
+            }
+            self.update_assignment(edge.source, final_ass)
+        }
+    }
+
+    /// Rising the value of the source vertex assignment if the target vertex was assigned
+    /// false in the last component assignment. If the source vertex already is true we
+    /// simply return. The return value is based on if a changed have been made.
+    fn process_negation(
+        &mut self,
+        edge: NegationEdge<V>,
+        ass_from_earlier: HashMap<V, bool>,
+    ) -> bool {
+        if *self.assignment.get(&edge.source).unwrap() {
+            false
+        } else {
+            let mut final_ass = true;
+
+            if *ass_from_earlier.get(&edge.target).unwrap() {
+                final_ass = false;
+            }
+
+            self.update_assignment(edge.source, final_ass)
+        }
```

Comment:
Pretty use the following will do the job
```rust
let new_ass = !*self.assignment.get(&edge.target).unwrap();
self.update_assignment(edge.source, new_ass)
```

Reply:
When implementing this, i have troubles that some cases enter a infinite loop.

Reply:
Nevermind, i think i misunderstood you comment.

Reply:
Yeah, the whole if-statement can be replace with the given fragment.

Reply:
I have played around with it, and i think we need the first check, since we are always raising the value of vertices. If we do not check the source then we can risk to lower it, depending of the target value.

Reply:
But the else branch can be replaced with the suggestion